lms/djangoapps/xblock_user_state/interface.py

```diff
@@ -0,0 +1,255 @@
+"""
+A baseclass for a generic client for accessing XBlock Scope.user_state field data.
+"""
+
+from abc import abstractmethod
+
+from contracts import contract, new_contract, ContractsMeta
+from opaque_keys.edx.keys import UsageKey
+from xblock.fields import Scope, ScopeBase
+
+new_contract('UsageKey', UsageKey)
+
+
+class XBlockUserStateClient(object):
+    """
+    First stab at an interface for accessing XBlock User State. This will have
+    use StudentModule as a backing store in the default case.
+
+    Scope/Goals:
+    1. Mediate access to all student-specific state stored by XBlocks.
+        a. This includes "preferences" and "user_info" (i.e. UserScope.ONE)
+        b. This includes XBlock Asides.
+        c. This may later include user_state_summary (i.e. UserScope.ALL).
+        d. This may include group state in the future.
+        e. This may include other key types + UserScope.ONE (e.g. Definition)
+    2. Assume network service semantics.
+        At some point, this will probably be calling out to an external service.
+        Even if it doesn't, we want to be able to implement circuit breakers, so
+        that a failure in StudentModule doesn't bring down the whole site.
+        This also implies that the client is running as a user, and whatever is
+        backing it is smart enough to do authorization checks.
+    3. This does not yet cover export-related functionality.
+
+    Open Questions:
+    1. Is it sufficient to just send the block_key in and extract course +
+       version info from it?
+    2. Do we want to use the username as the identifier? Privacy implications?
+       Ease of debugging?
+    3. Would a get_many_by_type() be useful?
+    """
+
+    __metaclass__ = ContractsMeta
+
+    class ServiceUnavailable(Exception):
+        """
+        This error is raised if the service backing this client is currently unavailable.
+        """
+        pass
+
+    class PermissionDenied(Exception):
+        """
+        This error is raised if the caller is not allowed to access the requested data.
+        """
+        pass
+
+    class DoesNotExist(Exception):
+        """
+        This error is raised if the caller has requested data that does not exist.
+        """
+        pass
+
+    @contract(
+        username="basestring",
+        block_key=UsageKey,
+        scope=ScopeBase,
+        fields="seq(basestring)|set(basestring)|None",
```

Comment:
Sadly, no, not that I could find. I think to implement that, it would have to actually inject a wrapper around the passed in argument to inspect the iterated values as they are delivered.